src/Content/Text/BBCode.php

```diff
@@ -1572,7 +1573,7 @@ function ($matches) use ($simple_html) {
 		}
 
 		// Replace non graphical smilies for external posts
-		if ($simple_html) {
+		if (!$nosmile && !$for_plaintext) {
```

Comment:
There is a massive difference between ```$simple_html``` and ```$for_plaintext```. I think that this change could be a problem for external connectors.

Reply:
Th way it was done:
- `BBCode::convert()` for displaying posts with `[nosmile]` in `prepare_text()`.
- `Smilies:replace(BBCode::convert())` for displaying posts without `[nosmile]` in `prepare_text()`.
- `BBCode::convert() { return Smilies:replace($text); }` for delivered messages (with `$simple_html`)

The new way is now
- `BBCode::convert()` for any post with `[nosmile]`
- `BBCode::convert() { return Smilies:replace($text); }` for any other post (with `$simple_html` or without) that is not intended for plaintext

In the end it is almost the same thing (`Smilies::replace` doesn't wrap `BBCode::convert` anymore), except that we were possibly replacing smilies in strings intended for plaintext.

Reply:
I still don't understand the change from ```$simple_html``` to ```$for_plaintext```.

Reply:
To explain it: The call ```$text = Smilies::replace($text);``` replaces all emojis with their unicode siblings. This has to be done for all connectors since they can't work with picture smileys.

Reply:
`Smilies::replace` replaces string emojis with whatever smilies are available. This only includes the unicode characters if the `unicode_emoji` addon is enabled, but it will also include all the basic smilie images, and the extended smilie images if the corresponding addon is enabled.

I'll explain further the removal of `$simple_html` while ignoring the `nosmile` condition.

### Before

We were using `prepare_text()` for local post display, and just `BBCode::convert($simple_html > 0)` for connector posts. Here's a simplified development of what each one was doing:
```php
function prepare_text($text)
{
    return Smilies::replace(BBCode::convert($text));
}
```
```php
function BBCode::convert($text, $simple_html > 0)
{
    // Escape pre-formatted blocks
    ...
    if ($simple_html) {
        $text = Smilies::convert($text);
    }
    ...
    // Unescape pre-formatted blocks
}
```
The problem was that `prepare_text()` was running `Smilies::replace` on the resulting HTML after the pre-formatted blocks were unescaped, and that `Smilies::replaceFromArray` was disregarding language code blocks.

### Intermediate

My first solution was to have `Smilies::convert` ran in a single place in `BBCode::convert` to take advantage of the code blocks escaping. This meant running `Smilies::convert` when `$simple_html` was also 0, which meant the initial condition wasn't warranted anymore, giving roughly:
```php
function prepare_text($text)
{
    return BBCode::convert($text);
}
```
```php
function BBCode::convert($text)
{
    // Escape pre-formatted blocks
    ...
    $text = Smilies::convert($text);
    ...
    // Unescape pre-formatted blocks
}
```

### Final

But then I realized that replacing smilies by image tags when `BBCode::convert` was used for plaintext wasn't warranted, it just doesn't make sense. So I added this condition that wasn't there before, which gave the impression I simply replaced `$simple_html` by `$for_plaintext` in the condition because the intermediate step in my reflexion was lost.
```php
function BBCode::convert($text)
{
    // Escape pre-formatted blocks
    ...
    if (!$for_plaintext) {
        $text = Smilies::convert($text);
    }
    ...
    // Unescape pre-formatted blocks
}
```